File: Server/Strategy.cs
using System;
using System.IO;
using System.Diagnostics;
using cAlgo.API;
using cAlgo.API.Internals;

namespace cAlgo.Robots;

public abstract class Strategy
{
    private readonly Api _api;
    private readonly Logger _logger;
    private readonly Robot _robot;

    private double? _lastPositionVolume;
    private double? _lastPositionStopLoss;
    private double? _lastPositionTakeProfit;
    private double? _lastPositionAskAboveTarget;
    private double? _lastPositionAskBelowTarget;
    private double? _lastPositionBidAboveTarget;
    private double? _lastPositionBidBelowTarget;

    private void CallShutdown() { _api.PackShutdown(); HandleCallback(); }
    private void CallComplete() { _api.PackComplete(); HandleCallback(); }
    private void CallAccount(IAccount account) { _api.PackAccount(account); HandleCallback(); }
    private void CallSymbol(Symbol symbol) { _api.PackSymbol(symbol); HandleCallback(); }
    private void CallOpenedBuy(Position position) { _api.PackOpenedBuy(position); HandleCallback(); }
    private void CallOpenedSell(Position position) { _api.PackOpenedSell(position); HandleCallback(); }
    private void CallModifiedBuyVolume(Position position) { _api.PackModifiedBuyVolume(position); HandleCallback(); }
    private void CallModifiedBuyStopLoss(Position position) { _api.PackModifiedBuyStopLoss(position); HandleCallback(); }
    private void CallModifiedBuyTakeProfit(Position position) { _api.PackModifiedBuyTakeProfit(position); HandleCallback(); }
    private void CallModifiedSellVolume(Position position) { _api.PackModifiedSellVolume(position); HandleCallback(); }
    private void CallModifiedSellStopLoss(Position position) { _api.PackModifiedSellStopLoss(position); HandleCallback(); }
    private void CallModifiedSellTakeProfit(Position position) { _api.PackModifiedSellTakeProfit(position); HandleCallback(); }
    private void CallClosedBuy(Position position) { _api.PackClosedBuy(position); HandleCallback(); }
    private void CallClosedSell(Position position) { _api.PackClosedSell(position); HandleCallback(); }
    private void CallBar(Bar bar) { _api.PackBar(bar); HandleCallback(); }
    private void CallAskAboveTarget(double ask) { _api.PackAskAboveTarget(ask); HandleCallback(); }
    private void CallAskBelowTarget(double ask) { _api.PackAskBelowTarget(ask); HandleCallback(); }
    private void CallBidAboveTarget(double bid) { _api.PackBidAboveTarget(bid); HandleCallback(); }
    private void CallBidBelowTarget(double bid) { _api.PackBidBelowTarget(bid); HandleCallback(); }

    private bool ClosePosition()
    {
        var position = _robot.Positions.Find(_robot.InstanceId);
        if (position != null && !_robot.ClosePosition(position).IsSuccessful) return false;
        _lastPositionVolume = null;
        _lastPositionStopLoss = null;
        _lastPositionTakeProfit = null;
        _lastPositionAskAboveTarget = null;
        _lastPositionAskBelowTarget = null;
        _lastPositionBidAboveTarget = null;
        _lastPositionBidBelowTarget = null;
        return true;
    }
    private void OpenPositionFixed(TradeType type, double volume, double? slPips, double? tpPips)
    {
        if (!ClosePosition()) return;
        var result = _robot.ExecuteMarketOrder(type, _robot.SymbolName, volume, _robot.InstanceId, slPips, tpPips);
        if (!result.IsSuccessful) return;
        _lastPositionVolume = result.Position.VolumeInUnits;
        _lastPositionStopLoss = result.Position.StopLoss;
        _lastPositionTakeProfit = result.Position.TakeProfit;
    }

    private void OpenPositionDynamic(TradeType type, double percentage, double slPips, double? tpPips)
    {
        var volume = _robot.Symbol.VolumeForProportionalRisk(ProportionalAmountType.Balance, percentage, slPips, RoundingMode.Down);
        OpenPositionFixed(type, volume, slPips, tpPips);
    }

    private void OnSignalBullishFixed(double volume, double? slPips, double? tpPips) { OpenPositionFixed(TradeType.Buy, volume, slPips, tpPips); }
    private void OnSignalBullishDynamic(double percentage, double slPips, double? tpPips) { OpenPositionDynamic(TradeType.Buy, percentage, slPips, tpPips); }
    private void OnSignalBearishFixed(double volume, double? slPips, double? tpPips) { OpenPositionFixed(TradeType.Sell, volume, slPips, tpPips); }
    private void OnSignalBearishDynamic(double percentage, double slPips, double? tpPips) { OpenPositionDynamic(TradeType.Sell, percentage, slPips, tpPips); }
    private void OnSignalSideways() { ClosePosition(); }

    private void OnModifyVolume(double percentage)
    {
        var position = _robot.Positions.Find(_robot.InstanceId);
        if (position == null) return;
        var volume = _robot.Symbol.NormalizeVolumeInUnits(position.VolumeInUnits * percentage / 100, RoundingMode.Down);
        _robot.ModifyPosition(position, volume);
    }

    private void OnModifyStopLoss(double? slPrice)
    {
        var position = _robot.Positions.Find(_robot.InstanceId);
        if (position == null) return;
        _robot.ModifyPosition(position, slPrice, position.TakeProfit);
    }

    private void OnModifyTakeProfit(double? tpPrice)
    {
        var position = _robot.Positions.Find(_robot.InstanceId);
        if (position == null) return;
        _robot.ModifyPosition(position, position.StopLoss, tpPrice);
    }

    private void HandleCallback()
    {
        var call = _api.UnpackHeader();
        double obligatoryVolume, obligatoryStopLoss;
        double? optionalStopLoss, optionalTakeProfit;
        switch (call)
        {
            case Api.IdReceive.Complete:
                break;
            case Api.IdReceive.SignalBullishFixed:
                (obligatoryVolume, optionalStopLoss, optionalTakeProfit) = _api.UnpackSignalFixed();
                OnSignalBullishFixed(obligatoryVolume, optionalStopLoss, optionalTakeProfit);
                break;
            case Api.IdReceive.SignalBullishDynamic:
                (obligatoryVolume, obligatoryStopLoss, optionalTakeProfit) = _api.UnpackSignalDynamic();
                OnSignalBullishDynamic(obligatoryVolume, obligatoryStopLoss, optionalTakeProfit);
                break;
            case Api.IdReceive.SignalSideways:
                OnSignalSideways();
                break;
            case Api.IdReceive.SignalBearishFixed:
                (obligatoryVolume, optionalStopLoss, optionalTakeProfit) = _api.UnpackSignalFixed();
                OnSignalBearishFixed(obligatoryVolume, optionalStopLoss, optionalTakeProfit);
                break;
            case Api.IdReceive.SignalBearishDynamic:
                (obligatoryVolume, obligatoryStopLoss, optionalTakeProfit) = _api.UnpackSignalDynamic();
                OnSignalBearishDynamic(obligatoryVolume, obligatoryStopLoss, optionalTakeProfit);
                break;
            case Api.IdReceive.ModifyVolume:
                OnModifyVolume(_api.UnpackObligatoryValue());
                break;
            case Api.IdReceive.ModifyStopLoss:
                OnModifyStopLoss(_api.UnpackOptionalValue());
                break;
            case Api.IdReceive.ModifyTakeProfit:
                OnModifyTakeProfit(_api.UnpackOptionalValue());
                break;
            case Api.IdReceive.AskAboveTarget:
                _lastPositionAskAboveTarget = _api.UnpackOptionalValue();
                break;
            case Api.IdReceive.AskBelowTarget:
                _lastPositionAskBelowTarget = _api.UnpackOptionalValue();
                break;
            case Api.IdReceive.BidAboveTarget:
                _lastPositionBidAboveTarget = _api.UnpackOptionalValue();
                break;
            case Api.IdReceive.BidBelowTarget:
                _lastPositionBidBelowTarget = _api.UnpackOptionalValue();
                break;
        }
    }

    protected Strategy(Robot robot, Logger.Verbose verbose)
    {
        _logger = new Logger(robot, verbose);
        _robot = robot;

        _robot.Positions.Opened += OnPositionOpened;
        _robot.Positions.Modified += OnPositionModified;
        _robot.Positions.Closed += OnPositionClosed;
        _robot.Bars.BarClosed += OnBar;
        _robot.Symbol.Tick += OnTick;

        _api = new Api(_robot.SymbolName, _robot.TimeFrame.Name, _logger);
        _api.Initialize();

        var baseDirectory = new DirectoryInfo(Environment.CurrentDirectory).Parent?.Parent?.Parent?.FullName;
        var scriptName = GetType().Name;
        var scriptPath = $@"{baseDirectory}\Sources\Client\{scriptName}.py";
        var scriptArgs = $"--verbose {verbose} --symbol {_robot.SymbolName} --timeframe {_robot.TimeFrame.Name}";
        var tabTitle = $"{scriptName} {_robot.SymbolName} {_robot.TimeFrame.Name}";
        var command = $"cmd.exe /k \"conda activate quant && python \"{scriptPath}\" {scriptArgs}\"";
        Process.Start("wt.exe", $"--window 0 new-tab --title \"{tabTitle}\" {command}");

        _api.Connect();

        CallAccount(_robot.Account);
        CallSymbol(_robot.Symbol);
        for (var i = 0; i < _robot.Bars.Count-1; i++) { CallBar(_robot.Bars[i]); }
        CallComplete();
    }

    private void OnPositionOpened(PositionOpenedEventArgs args)
    {
        if (!string.Equals(args.Position.Label, _robot.InstanceId)) return;
        if (args.Position.TradeType == TradeType.Buy) CallOpenedBuy(args.Position); else CallOpenedSell(args.Position);
    }

    private void OnPositionModified(PositionModifiedEventArgs args)
    {
        if (!string.Equals(args.Position.Label, _robot.InstanceId)) return;
        if (_lastPositionVolume != null && Math.Abs(args.Position.VolumeInUnits - (double)_lastPositionVolume) > double.Epsilon)
        {
            _lastPositionVolume = args.Position.VolumeInUnits;
            if (args.Position.TradeType == TradeType.Buy) CallModifiedBuyVolume(args.Position); else CallModifiedSellVolume(args.Position);
            return;
        }
        if ((_lastPositionStopLoss == null && args.Position.StopLoss != null) ||
            (_lastPositionStopLoss != null && args.Position.StopLoss == null) ||
            (_lastPositionStopLoss != null && args.Position.StopLoss != null && Math.Abs((double)args.Position.StopLoss - (double)_lastPositionStopLoss) > double.Epsilon))
        {
            _lastPositionStopLoss = args.Position.StopLoss;
            if (args.Position.TradeType == TradeType.Buy) CallModifiedBuyStopLoss(args.Position); else CallModifiedSellStopLoss(args.Position);
            return;
        }
        if ((_lastPositionTakeProfit == null && args.Position.TakeProfit != null) ||
            (_lastPositionTakeProfit != null && args.Position.TakeProfit == null) ||
            (_lastPositionTakeProfit != null && args.Position.TakeProfit != null && Math.Abs((double)args.Position.TakeProfit - (double)_lastPositionTakeProfit) > double.Epsilon))
        {
            _lastPositionTakeProfit = args.Position.TakeProfit;
            if (args.Position.TradeType == TradeType.Buy) CallModifiedBuyTakeProfit(args.Position); else CallModifiedSellTakeProfit(args.Position);
        }
    }

    private void OnPositionClosed(PositionClosedEventArgs args)
    {
        if (!string.Equals(args.Position.Label, _robot.InstanceId)) return;
        if (args.Position.TradeType == TradeType.Buy) CallClosedBuy(args.Position); else CallClosedSell(args.Position);
    }

    private void OnBar(BarClosedEventArgs args) { CallBar(args.Bars.LastBar); }

    private void OnTick(SymbolTickEventArgs args)
    {
        if (_lastPositionAskAboveTarget != null && args.Ask >= _lastPositionAskAboveTarget) { CallAskAboveTarget(args.Ask); _lastPositionAskAboveTarget = null; }
        if (_lastPositionAskBelowTarget != null && args.Ask <= _lastPositionAskBelowTarget) { CallAskBelowTarget(args.Ask); _lastPositionAskBelowTarget = null; }
        if (_lastPositionBidAboveTarget != null && args.Bid >= _lastPositionBidAboveTarget) { CallBidAboveTarget(args.Bid); _lastPositionBidAboveTarget = null; }
        if (_lastPositionBidBelowTarget != null && args.Bid <= _lastPositionBidBelowTarget) { CallBidBelowTarget(args.Bid); _lastPositionBidBelowTarget = null; }
    }

    public void OnError(Error error)
    {
        _logger.Error("An unexpected error occured in the server execution");
        _logger.Error(error.TradeResult.ToString());
        _robot.Stop();
    }

    public void OnException(Exception exception)
    {
        _logger.Error("An unexpected exception occured in the server execution");
        _logger.Error(exception.ToString());
        _robot.Stop();
    }

    public void OnShutdown()
    {
        _logger.Warning("Shutdown strategy and safely terminate operations");
        CallShutdown();
        _api.Disconnect();
    }
}

File: Client/Strategy/Strategy.py
from abc import ABC

from .Api import API, IdReceive, IdSend
from .Machine import Machine


class Strategy(ABC):

    def __init__(self, symbol, timeframe, logger):
        self.symbol = symbol
        self.timeframe = timeframe
        self.logger = logger

        self.signal_machine: Machine = self.create_signal_management()
        self.risk_machine: Machine = self.create_risk_management()

    def run(self):
        with API(self.symbol, self.timeframe, self.logger) as self.api:
            while not (self.risk_machine.at.end and self.signal_machine.at.end):
                call = self.api.unpack_header()
                match call:
                    case IdReceive.Shutdown.value:
                        self.logger.warning("Shutdown strategy and safely terminate operations")
                        callback, *callback_args = self.__call_shutdown()
                    case IdReceive.Complete.value:
                        callback, *callback_args = self.__call_complete()
                    case IdReceive.Account.value:
                        callback, *callback_args = self.__call_account(*self.api.unpack_account())
                    case IdReceive.Symbol.value:
                        callback, *callback_args = self.__call_symbol(*self.api.unpack_symbol())
                    case IdReceive.OpenedBuy.value:
                        callback, *callback_args = self.__call_opened_buy(*self.api.unpack_position())
                    case IdReceive.OpenedSell.value:
                        callback, *callback_args = self.__call_opened_sell(*self.api.unpack_position())
                    case IdReceive.ModifiedBuyVolume.value:
                        callback, *callback_args = self.__call_modified_buy_volume(*self.api.unpack_position())
                    case IdReceive.ModifiedBuyStopLoss.value:
                        callback, *callback_args = self.__call_modified_buy_stop_loss(*self.api.unpack_position())
                    case IdReceive.ModifiedBuyTakeProfit.value:
                        callback, *callback_args = self.__call_modified_buy_take_profit(*self.api.unpack_position())
                    case IdReceive.ModifiedSellVolume.value:
                        callback, *callback_args = self.__call_modified_sell_volume(*self.api.unpack_position())
                    case IdReceive.ModifiedSellStopLoss.value:
                        callback, *callback_args = self.__call_modified_sell_stop_loss(*self.api.unpack_position())
                    case IdReceive.ModifiedSellTakeProfit.value:
                        callback, *callback_args = self.__call_modified_sell_take_profit(*self.api.unpack_position())
                    case IdReceive.ClosedBuy.value:
                        callback, *callback_args = self.__call_closed_buy(*self.api.unpack_position())
                    case IdReceive.ClosedSell.value:
                        callback, *callback_args = self.__call_closed_sell(*self.api.unpack_position())
                    case IdReceive.Bar.value:
                        callback, *callback_args = self.__call_bar(*self.api.unpack_bar())
                    case IdReceive.AskAboveTarget.value:
                        callback, *callback_args = self.__call_ask_above_target(self.api.unpack_target())
                    case IdReceive.AskBelowTarget.value:
                        callback, *callback_args = self.__call_ask_below_target(self.api.unpack_target())
                    case IdReceive.BidAboveTarget.value:
                        callback, *callback_args = self.__call_bid_above_target(self.api.unpack_target())
                    case IdReceive.BidBelowTarget.value:
                        callback, *callback_args = self.__call_bid_below_target(self.api.unpack_target())
                match callback:
                    case IdSend.Complete.value:
                        self.api.pack_complete()
                    case IdSend.SignalBullishFixed.value:
                        self.api.pack_signal_bullish_fixed(*callback_args)
                    case IdSend.SignalBullishDynamic.value:
                        self.api.pack_signal_bullish_dynamic(*callback_args)
                    case IdSend.SignalSideways.value:
                        self.api.pack_signal_sideways()
                    case IdSend.SignalBearishFixed.value:
                        self.api.pack_signal_bearish_fixed(*callback_args)
                    case IdSend.SignalBearishDynamic.value:
                        self.api.pack_signal_bearish_dynamic(*callback_args)
                    case IdSend.ModifyVolume.value:
                        self.api.pack_modify_volume(*callback_args)
                    case IdSend.ModifyStopLoss.value:
                        self.api.pack_modify_stop_loss(*callback_args)
                    case IdSend.ModifyTakeProfit.value:
                        self.api.pack_modify_take_profit(*callback_args)
                    case IdSend.AskAboveTarget.value:
                        self.api.pack_ask_above_target(*callback_args)
                    case IdSend.AskBelowTarget.value:
                        self.api.pack_ask_below_target(*callback_args)
                    case IdSend.BidAboveTarget.value:
                        self.api.pack_bid_above_target(*callback_args)
                    case IdSend.BidBelowTarget.value:
                        self.api.pack_bid_below_target(*callback_args)

    def __create_dummy_machine(self):
        machine = Machine(None, self.symbol, self.timeframe, self.logger)
        machine.create_state(None, True)
        return machine

    def create_risk_management(self):
        return self.__create_dummy_machine()

    def create_signal_management(self):
        return self.__create_dummy_machine()

    @staticmethod
    def __call(signal_call, risk_call, *call_args):
        signal_return = signal_call(*call_args)
        risk_return = risk_call(*call_args)
        if signal_return is not None:
            callback, *callback_args = signal_return
            return callback, *callback_args
        if risk_return is not None:
            callback, *callback_args = risk_return
            return callback, *callback_args
        return IdSend.Complete.value, None

    def __call_shutdown(self):
        return self.__call(self.signal_machine.call_shutdown, self.risk_machine.call_shutdown)

    def __call_complete(self):
        return self.__call(self.signal_machine.call_complete, self.risk_machine.call_complete)

    def __call_account(self, *account):
        return self.__call(self.signal_machine.call_account, self.risk_machine.call_account, *account)

    def __call_symbol(self, *symbol):
        return self.__call(self.signal_machine.call_symbol, self.risk_machine.call_symbol, *symbol)

    def __call_opened_buy(self, *position):
        return self.__call(self.signal_machine.call_opened_buy, self.risk_machine.call_opened_buy, *position)

    def __call_opened_sell(self, *position):
        return self.__call(self.signal_machine.call_opened_sell, self.risk_machine.call_opened_sell, *position)

    def __call_modified_buy_volume(self, *position):
        return self.__call(self.signal_machine.call_modified_buy_volume, self.risk_machine.call_modified_buy_volume, *position)

    def __call_modified_buy_stop_loss(self, *position):
        return self.__call(self.signal_machine.call_modified_buy_stop_loss, self.risk_machine.call_modified_buy_stop_loss, *position)

    def __call_modified_buy_take_profit(self, *position):
        return self.__call(self.signal_machine.call_modified_buy_take_profit, self.risk_machine.call_modified_buy_take_profit, *position)

    def __call_modified_sell_volume(self, *position):
        return self.__call(self.signal_machine.call_modified_sell_volume, self.risk_machine.call_modified_sell_volume, *position)

    def __call_modified_sell_stop_loss(self, *position):
        return self.__call(self.signal_machine.call_modified_sell_stop_loss, self.risk_machine.call_modified_sell_stop_loss, *position)

    def __call_modified_sell_take_profit(self, *position):
        return self.__call(self.signal_machine.call_modified_sell_take_profit, self.risk_machine.call_modified_sell_take_profit, *position)

    def __call_closed_buy(self, *position):
        return self.__call(self.signal_machine.call_closed_buy, self.risk_machine.call_closed_buy, *position)

    def __call_closed_sell(self, *position):
        return self.__call(self.signal_machine.call_closed_sell, self.risk_machine.call_closed_sell, *position)

    def __call_bar(self, *bar):
        return self.__call(self.signal_machine.call_bar, self.risk_machine.call_bar, *bar)

    def __call_ask_above_target(self, *target):
        return self.__call(self.signal_machine.call_ask_above_target, self.risk_machine.call_ask_above_target, *target)

    def __call_ask_below_target(self, *target):
        return self.__call(self.signal_machine.call_ask_below_target, self.risk_machine.call_ask_below_target, *target)

    def __call_bid_above_target(self, *target):
        return self.__call(self.signal_machine.call_bid_above_target, self.risk_machine.call_bid_above_target, *target)

    def __call_bid_below_target(self, *target):
        return self.__call(self.signal_machine.call_bid_below_target, self.risk_machine.call_bid_below_target, *target)
